Upgrade TypeScript to v4.9.3

aaronccasanova opened the pull request:
### WHAT is this pull request doing?

This PR upgrades the root TypeScript version to `v4.9.3` and performs minimal changes to address build failures.

TODO:
- [x] Create snapshot release
- [ ] Tophat the Admin

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

<!--
  Give as much information as needed to experiment with the component
  in the playground.
-->

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';
import {Page} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      {/* Add the code you want to test in here */}
    </Page>
  );
}
```

</details>

### 🎩 checklist

- [ ] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [ ] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

aaronccasanova committed:
Upgrade TypeScript to v4.8.3

aaronccasanova committed:
Revert type parameter rename to reduce overall changes

aaronccasanova committed:
docs: Add changeset entry

aaronccasanova committed:
Update polaris-for-figma ts-loader version

sam-b-rose commented on polaris-cli/package.json:
```diff
@@ -26,9 +26,7 @@
-  "devDependencies": {
-    "typescript": "^4.7.4"
-  },
+  "devDependencies": {},
```
We should just remove `devDeps` if there isn't anything included 💥

sam-b-rose reviewed (dismissed):
Thank you for keeping this updated ⬆️ 👍  looks good so far

aaronccasanova committed:
Merge branch 'main' of https://github.com/Shopify/polaris into upgrade-typescript

aaronccasanova committed:
Update changeset entry

laurkim committed:
[polaris.shopify.com] Add content for uppercase typography (#8235)

### WHY are these changes introduced?

Added a section to Design/Typography in the style guide to call out that
the design system no longer supports uppercase typography styles.

### WHAT is this pull request doing?

<img width="1353" alt="01-30-evh8p-8m14u"
src="https://user-images.githubusercontent.com/26749317/216118646-fafc6bda-4044-4e94-a25e-6157ac361485.png">

### How to 🎩

🖥 [Local development
instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting
guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog
guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

### 🎩 checklist

- [ ] Tested on
[mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [ ] Tested on [multiple
browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [ ] Tested for
[accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [x] [Tophatted
documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md)
changes in the style guide

---------

Co-authored-by: aveline <aveline@users.noreply.github.com>

alex-page committed:
Merge branch 'main' into v11-major

alex-page committed:
Merge branch 'main' into upgrade-typescript

alex-page committed:
Merge branch 'v11-major' into upgrade-typescript

alex-page committed:
Merge branch 'rm-dldts' into upgrade-typescript

alex-page committed:
Remove changesets
